### PR TITLE
PB-7332 : API Changes for Sharing / Unsharing Clusters to Users and Groups

### PR DESCRIPTION
**Pull request by pallav-px**
Share Cluster : Share Cluster to users and groups
Unshare Cluster: Revoke Shared Clusters from users and groups

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** (optional)
Closes #

**Special notes for your reviewer**:

### CODE REVIEW

File: pkg/apis/v1/api.proto [hunk 1]
```diff
@@ -1188,6 +1188,22 @@ service Cluster {
             body : "*"
         };
     }
+
+    // share cluster add share access of your cluster to another users
```
Comment (vsundarraj-px) [inline pkg/apis/v1/api.proto:1192]: nit: the first word is the functionName itself like
//ShareClsuter shares access to the cluster to the user(s) or group(s).

File: pkg/apis/v1/api.proto [hunk 2]
```diff
@@ -1188,6 +1188,22 @@ service Cluster {
...
+
+    // share cluster add share access of your cluster to another users 
+    rpc ShareCluster(ShareClusterRequest) returns (ShareClusterResponse) {
+        option (google.api.http) = {
+            patch : "/v1/sharecluster"
+            body : "*"
+        };
+    }
+
+    // Unshare cluster remove sharing access of your cluster from another users
```
Comment (vsundarraj-px) [inline pkg/apis/v1/api.proto:1200]: //unshareClsuter removes shared access for the user(s) or group(s) to the cluster

File: pkg/apis/v1/api.proto [hunk 3]
```diff
@@ -1302,6 +1318,31 @@ message ClusterDeleteRequest {
 // Define ClusterInspectResponse struct
 message ClusterDeleteResponse {}
 
+
+message ShareClusterRequest{
+    string org_id = 1;
+    ObjectRef cluster_ref = 2;
+    // it will share the cluster to all the user and group
```
Comment (vsundarraj-px) [inline pkg/apis/v1/api.proto:1325]: //userid of the user to share the cluster with

File: pkg/apis/v1/api.proto [hunk 4]
```diff
@@ -1302,6 +1318,31 @@ message ClusterDeleteRequest {
...
+    repeated string user = 3;
+    repeated string group = 4;
```
Comment (vsundarraj-px) [inline pkg/apis/v1/api.proto:1327]: //groupIDs to share the cluster with

File: pkg/apis/v1/api.proto [hunk 5]
```diff
@@ -1302,6 +1318,31 @@ message ClusterDeleteRequest {
...
+    bool share_cluster_backup = 5;
+}
+
+message ShareClusterResponse{}
+
+message UnShareClusterRequest{
+    string org_id = 1;
+    ObjectRef cluster_ref = 2;
+    // it will unshare the cluster from all the user and group
+    repeated string user = 3;
```
Comment (vsundarraj-px) [inline pkg/apis/v1/api.proto:1338]: Same as above

File: pkg/apis/v1/api.proto [hunk 6]
```diff
@@ -1302,6 +1318,31 @@ message ClusterDeleteRequest {
...
 message ClusterDeleteResponse {}
 
+
+message ShareClusterRequest{
+    string org_id = 1;
+    ObjectRef cluster_ref = 2;
+    // it will share the cluster to all the user and group
+    repeated string user = 3;
+    repeated string group = 4;
+    // share cluster backup is optional, if set to true, it will share the cluster backup
```
Comment (vsundarraj-px) [inline pkg/apis/v1/api.proto:1328]: // If set to true will additionally share existing backups

File: pkg/apis/v1/api.proto [hunk 7]
```diff
@@ -1302,6 +1318,31 @@ message ClusterDeleteRequest {
...
+
+message ShareClusterResponse{}
+
+message UnShareClusterRequest{
+    string org_id = 1;
+    ObjectRef cluster_ref = 2;
+    // it will unshare the cluster from all the user and group
+    repeated string user = 3;
+    repeated string group = 4;
+    // unshare cluster backup is optional, if set to true, it will unshare the cluster backup
```
Comment (vsundarraj-px) [inline pkg/apis/v1/api.proto:1340]: // if set will unshare shared backups.

File: pkg/apis/v1/api.proto [hunk 8]
```diff
@@ -1188,6 +1188,22 @@ service Cluster {
             body : "*"
         };
     }
+
+    // ShareClsuter shares access to the cluster to the user(s) or group(s)
```
Comment (vsundarraj-px) [inline pkg/apis/v1/api.proto:1192]: type: ShareClsuter -> ShareCluster, cannot ignore these comments cos they will show up as help in swagger

File: pkg/apis/v1/api.proto [hunk 9]
```diff
@@ -1302,6 +1318,35 @@ message ClusterDeleteRequest {
...
+
+message UnShareClusterRequest{
+    string org_id = 1;
+    ObjectRef cluster_ref = 2;
+    // userid of the user to share the cluster with
+    repeated string user = 3;
+    // group to share the cluster with
+    repeated string group = 4;
+    // unshare_cluster_backup is optional, if set to true, will unshare shared backups.
+
```
Comment (vsundarraj-px) [inline pkg/apis/v1/api.proto:1343]: nit: Extra lines

**Review by vsundarraj-px — APPROVED** [review]
LGTM

File: pkg/apis/v1/api.proto [hunk 10]
```diff
@@ -1188,6 +1188,22 @@ service Cluster {
             body : "*"
         };
     }
+
+    // ShareCluster shares access to the cluster to the user(s) or group(s)
+    rpc ShareCluster(ShareClusterRequest) returns (ShareClusterResponse) {
```
Comment (prashanthpx) [inline pkg/apis/v1/api.proto:1193]: For existing RBAC object we have `UpdateOwnership` and cluster Share is extension of existing RBAC to clusterObject. Any reason we are introducing two new API?

File: pkg/apis/v1/api.proto [hunk 11]
```diff
@@ -1304,6 +1320,31 @@ message ClusterDeleteRequest {
 // Define ClusterInspectResponse struct
 message ClusterDeleteResponse {}
 
+message ShareClusterRequest{
+    string org_id = 1;
+    ObjectRef cluster_ref = 2;
+    // userid of the user to share the cluster with
+    repeated string user = 3;
```
Comment (vsundarraj-px) [inline pkg/apis/v1/api.proto:1327]: users since its an array

File: pkg/apis/v1/api.proto (hunk 4 above)
Comment (vsundarraj-px) [reply pkg/apis/v1/api.proto:1329]: can we use groups since its an array?

**Pull request merged**